Readme.md
# Nodealytics

(Formerly Googalytics)

Simple node module to send custom server-side events to Google Analytics

Heavily influenced by the Gabba project from the guys at the Hybrid Group (http://github.com/hybridgroup/gabba)

## Examples

### Initialize

```javascript
var NA = require("nodealytics");
NA.initialize('UA-12345678-1', 'someplace.com', function () {
  NA.trackPage('page title', 'pageName', function (err, resp) {
    if (!err, resp.statusCode === 200) {
      console.log('it worked!');
    }
  });
});
```

### What do those variables mean?!

The answers, as per usual, can be learned from the Google: https://developers.google.com/analytics/resources/articles/gaTrackingTroubleshooting#gifParameters


CHANGELOG
v0.0.5 / 2013-04-18
--------------------

* NA.trackPage now supports custom params

v0.0.4 / 2013-03-13
--------------------

* Remove underscore dependency
* Improved Google Analytics UA id string validation
* Make callback argument to `runQuery` function optional



tests/run.js
var GA = require("googalytics");
var should = require("should");
var assert = require("assert");

describe('just get started', function (done) {
  (1).should.eql(1);
  done();
});

tests/test.js
var NA = require("../lib/main.js");
var should = require("should");

describe('initialize', function () {
  it('should require an account ID', function (done) {
    NA.initialize('', '', function (err, resp) {
      err.message.should.eql('Account ID is invalid');
      done();
    });
  });

  it('should require a valid account ID', function (done) {
    NA.initialize('AB-1230404-1', 'testsite.com', function (err, resp) {
      err.message.should.eql('Account ID is invalid');
      done();
    });
  });

  it('should accept an account ID without profile number', function (done) {
    NA.initialize('UA-12345678', 'testsite.com', done);
  });

  it('should require a domain', function (done) {
    NA.initialize('UA-12345678-1', '', function (err, resp) {
      err.message.should.eql('Domain is invalid');
      done();
    });
  });

  it('should initialize with the appropriate parameters', function (done) {
    NA.initialize('UA-33709401-1', 'testsite.com', done);
  });
});

describe('page views', function () {
  it('should register a page view', function (done){
    NA.trackPage('some title', 'some page', function (err, resp) {
      resp.statusCode.should.eql(200);
      done();
    });
  });

  it('should register a page view with custom gif params', function (done){
    NA.trackPage('some title', 'some page', {utmr: 'http://github.com'}, function (err, resp) {
      resp.statusCode.should.eql(200);
      resp.request.path.should.match(/&utmr=http:\/\/github.com/);
      done();
    });
  });
});

describe('events', function () {
  it('should register an event', function (done) {
    NA.trackEvent('test event', 'boom', function (err, resp) {
      resp.statusCode.should.eql(200);
      done();
    });
  });
});

describe('mapParams()', function () {
  it('should stringify params correctly', function () {
    var params = {
      'param1': 'valueA',
      'param2': 'valueB',
      'param3': 'valueC'
    }
    var result = NA.mapParams(params);
    var expected = 'param1=valueA&param2=valueB&param3=valueC';

    result.should.eql(expected);
  });
});



lib/googalytics.js
var request = require('request');
var _ = require('underscore')._;

var randomId = function (seed) {
  if (seed) {
    return Math.floor(Math.random() * seed + 1000000000);
  }

  return Math.floor(Math.random() * 8999999999 + 1000000000);
};

var cookieParams = function () {
  var utma1 = randomId();
  var utma2 = randomId();
  var today = Date.now();

  var str1 = '1.' + utma1 + '00145214523.' + utma2 + '.' + today + '.' + today + '.15';
  
  return '__utma=' + str1 + ';+__utmz=1.' + today + '.1.1.utmcsr=(direct)|utmccn=(direct)|utmcmd=(none);';
};


var eventData = function (category, action, label, value) {
  var data = '5(' + category + '*' + action;
  if (label) {
    data += '*' + label + ')';
  } else {
    data += ')';
  }
  if (value) {
    data += '(' + value + ')';
  }
  return data;
};

var GOOGLE_HOST = 'www.google-analytics.com';
var BEACON_PATH = '/__utm.gif';
var USER_AGENT = 'Googalytics Agent';

var options = {
  utmwv : "4.4sh", // GA version
  utmcs : "UTF-8", // charset
  utmul : "en-us", // language

  utmn : randomId(),
  utmhid : randomId(),
  user_agent : USER_AGENT,

  custom_vars : [] // TODO: Implement custom variables
};


exports.initialize = function(accountId, domain, callback) {
  options.utmac = accountId;
  options.utmhn = domain;

  return callback();
};


exports.trackPage = function (title, page, utmhid, callback) {
  if (typeof utmhid === 'function') {
    callback = utmhid;
    utmhid = undefined;
  }

  if (typeof options.utmac === 'undefined') {
    return new Error('Google Analytics Account ID is undefined');
  }
  if (typeof options.utmhn === 'undefined') {
    return new Error('Google Analytics Domain is undefined');
  }
  
  var pageViewParams = {
    utmwv : options.utmwv,
    utmn : options.utmn,
    utmhn : options.utmhn,
    utmcs : options.utmcs,
    utmul : options.utmul,
    utmdt : title,
    utmhid : ( utmhid || randomId() ),
    utmp : page,
    utmac : options.utmac,
    utmcc : cookieParams()
  };

  return this.runQuery(pageViewParams, callback);
};

exports.trackEvent = function (category, action, label, value, utmni, utmhid, callback) {
  if (typeof label === 'function') {
    callback = label;
    label = undefined;
  } else if (typeof value === 'function') {
    callback = value;
    value = undefined;
  } else if (typeof utmni === 'function') {
    callback = utmni;
    utmni = undefined;
  } else if (typeof utmhid === 'function') {
    callback = utmhid;
    utmhid = undefined;
  }

  if (typeof options.utmac === 'undefined') {
    return callback(new Error('Google Analytics Account ID is undefined'));
  }
  if (typeof options.utmhn === 'undefined') {
    return callback(new Error('Google Analytics Domain is undefined'));
  }

  if (typeof utmni !== 'undefined' && typeof utmni !== 'boolean') {
    return callback(new Error('utmni must be a boolean'));
  }

  var eventParams = {
    utmwv : options.utmwv,
    utmn : options.utmn,
    utmhn : options.utmhn,
    utmt : 'event',
    utme : eventData(category, action, label, value), // #{custom_var_data}, TODO: work on custom data
    utmcs : options.utmcs,
    utmul : options.utmul,
    utmhid : (utmhid || randomId()),
    utmac : options.utmac,
    utmcc : options.utmcc || cookieParams()
  };

  if (utmni) {
    eventParams.utmni = 1; // 1 for non interactive event, excluded from bounce calcs
  }

  return this.runQuery(eventParams, callback);
};

exports.runQuery = function (params, callback) {
  
  var query = 'http://' + GOOGLE_HOST + BEACON_PATH + '?';

  _.each(params, function (val, key) {
    query += key + '=' + val + '&';
  });
  query = query.slice(0,-1);

  request.get(query, function (error, response, body) {
    if (!error && response.statusCode == 200) {
      // console.log(body); 
    }
    return callback(error, response);
  });

};

lib/main.js
var request = require('request');

var randomId = function (seed) {
  if (seed) {
    return Math.floor(Math.random() * seed + 1000000000);
  }

  return Math.floor(Math.random() * 8999999999 + 1000000000);
};

var cookieParams = function () {
  var utma1 = randomId();
  var utma2 = randomId();
  var today = Date.now();

  var str1 = '1.' + utma1 + '00145214523.' + utma2 + '.' + today + '.' + today + '.15';
  
  return '__utma=' + str1 + ';+__utmz=1.' + today + '.1.1.utmcsr=(direct)|utmccn=(direct)|utmcmd=(none);';
};


var eventData = function (category, action, label, value) {
  var data = '5(' + category + '*' + action;
  if (label) {
    data += '*' + label + ')';
  } else {
    data += ')';
  }
  if (value) {
    data += '(' + value + ')';
  }
  return data;
};

var GOOGLE_HOST = 'www.google-analytics.com';
var BEACON_PATH = '/__utm.gif';
var USER_AGENT = 'Nodealytics Agent';

var options = {
  utmwv : "4.4sh", // Google Analytics version
  utmcs : "UTF-8", // charset
  utmul : "en-us", // language

  utmn : randomId(),
  utmhid : randomId(),
  user_agent : USER_AGENT,

  custom_vars : [] // TODO: Implement custom variables
};


exports.initialize = function(accountId, domain, callback) {
  options.utmac = accountId;
  if (!/UA-[0-9-]+[0-9]+/i.test(accountId)) {
    return callback(new Error('Account ID is invalid'));
  }

  if (domain.length < 1) {
    return callback(new Error('Domain is invalid'));
  }

  options.utmhn = domain;

  if (callback) {
    return callback();
  }
};


exports.trackPage = function (title, page, params, callback) {
  if (typeof params === 'function') {
    callback = params;
    params = {};
  }

  if (typeof options.utmac === 'undefined') {
    return new Error('Google Analytics Account ID is undefined');
  }
  if (typeof options.utmhn === 'undefined') {
    return new Error('Google Analytics Domain is undefined');
  }

  var pageViewParams = {
    utmwv : options.utmwv,
    utmn : options.utmn,
    utmhn : options.utmhn,
    utmcs : options.utmcs,
    utmul : options.utmul,
    utmdt : title,
    utmhid : randomId(),
    utmp : page,
    utmac : options.utmac,
    utmcc : cookieParams()
  };

  Object.keys(params).forEach(function (key, i) {
    pageViewParams[key] = params[key];
  });

  return this.runQuery(pageViewParams, callback);
};

exports.trackEvent = function (category, action, label, value, utmni, utmhid, callback) {
  if (typeof label === 'function') {
    callback = label;
    label = undefined;
  } else if (typeof value === 'function') {
    callback = value;
    value = undefined;
  } else if (typeof utmni === 'function') {
    callback = utmni;
    utmni = undefined;
  } else if (typeof utmhid === 'function') {
    callback = utmhid;
    utmhid = undefined;
  }

  if (typeof options.utmac === 'undefined') {
    return callback(new Error('Google Analytics Account ID is undefined'));
  }
  if (typeof options.utmhn === 'undefined') {
    return callback(new Error('Google Analytics Domain is undefined'));
  }

  if (typeof utmni !== 'undefined' && typeof utmni !== 'boolean') {
    return callback(new Error('utmni must be a boolean'));
  }

  var eventParams = {
    utmwv : options.utmwv,
    utmn : options.utmn,
    utmhn : options.utmhn,
    utmt : 'event',
    utme : eventData(category, action, label, value), // #{custom_var_data}, TODO: work on custom data
    utmcs : options.utmcs,
    utmul : options.utmul,
    utmhid : (utmhid || randomId()),
    utmac : options.utmac,
    utmcc : options.utmcc || cookieParams()
  };

  if (utmni) {
    eventParams.utmni = 1; // 1 for non interactive event, excluded from bounce calcs
  }

  return this.runQuery(eventParams, callback);
};

exports.mapParams = function(params) {
  var result = '';

  Object.keys(params).forEach(function (key, i) {
    result += key + '=' + params[key] + '&';
  });
  result = result.slice(0, -1);

  return result;
};

exports.runQuery = function (params, callback) {
  var query = 'http://' + GOOGLE_HOST + BEACON_PATH + '?';

  query += exports.mapParams(params);

  request.get(query, function (error, response, body) {
    if (!error && response.statusCode == 200) {
      // console.log(body);
    }
    if (callback) {
      return callback(error, response);
    }
  });

};

exports.VERSION = 0.0.5;

